File: scripts/seed_investors.py
import pandas as pd
from sqlalchemy import create_engine, text
import os
import re
from dotenv import load_dotenv

# Load environment variables from .env.local
load_dotenv('.env.local')

# Database connection URL from environment variable
DB_URL = os.getenv('DATABASE_URL')

if not DB_URL:
    print("Error: DATABASE_URL not found in environment variables.")
    print("Please set DATABASE_URL in your .env.local file.")
    print("Format: postgresql://postgres:[PASSWORD]@[HOST]:5432/postgres")
    exit(1)

def clean_currency(value):
    if pd.isna(value) or value == '':
        return None
    # Remove $, commas, and whitespace
    clean_val = re.sub(r'[$,\s]', '', str(value))
    try:
        return float(clean_val)
    except ValueError:
        return None

def seed():
    engine = create_engine(DB_URL)
    csv_path = "public/investors.csv"
    
    if not os.path.exists(csv_path):
        print(f"Error: {csv_path} not found!")
        return

    print(f"Reading CSV from {csv_path}...")
    df = pd.read_csv(csv_path)
    
    # Mapping CSV columns to Database columns
    mapping = {
        'Investor name': 'firm_name',
        'Website': 'website',
        'Global HQ': 'hq_location',
        'Countries of investment': 'investment_focus',
        'Stage of investment': 'stages',
        'Investment thesis': 'investment_thesis',
        'Investor type': 'investor_type',
        'First cheque minimum': 'min_check_size',
        'First cheque maximum': 'max_check_size'
    }
    
    df = df.rename(columns=mapping)
    
    # Cleaning
    print("Normalizing data...")
    df = df.drop_duplicates(subset=['firm_name', 'website'], keep='first')
    df['min_check_size'] = df['min_check_size'].apply(clean_currency)
    df['max_check_size'] = df['max_check_size'].apply(clean_currency)
    
    # Trim strings and handle nulls
    for col in df.select_dtypes(include=['object']).columns:
        df[col] = df[col].str.strip().fillna('')

    print(f"Inserting {len(df)} records into 'european_investors'...")
    # Using method='multi' for faster batch insertion if supported, 
    # but default to_sql is usually fine for 1MB.
    df.to_sql('european_investors', engine, if_exists='append', index=False)
    
    with engine.connect() as conn:
        result = conn.execute(text("SELECT COUNT(*) FROM european_investors"))
        count = result.scalar()
        print(f"Success! Total records in 'european_investors': {count}")

if __name__ == "__main__":
    seed()

File: scripts/seed_investors.mjs
import fs from 'fs';
import csv from 'csv-parser';
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';
import path from 'path';

dotenv.config({ path: '.env.local' });

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const supabaseKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseKey) {
    console.error('Missing Supabase credentials in .env.local');
    process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

const csvPath = path.resolve('public/investors.csv');

function cleanCurrency(value) {
    if (!value) return null;
    const cleanVal = value.replace(/[$,\s]/g, '');
    const num = parseFloat(cleanVal);
    return isNaN(num) ? null : num;
}

const results = [];

console.log(`Reading CSV from ${csvPath}...`);

let firstRow = true;
fs.createReadStream(csvPath)
    .pipe(csv())
    .on('data', (data) => {
        // Handle BOM in 'Investor name' or find the key that ends with 'Investor name'
        const nameKey = Object.keys(data).find(k => k.includes('Investor name'));
        const firmName = data[nameKey]?.trim();
        if (!firmName) return;

        // Mapping CSV columns to Database columns
        const mapped = {
            firm_name: firmName,
            website: data['Website']?.trim() || '',
            hq_location: data['Global HQ']?.trim() || '',
            investment_focus: data['Countries of investment']?.trim() || '',
            stages: data['Stage of investment']?.trim() || '',
            investment_thesis: data['Investment thesis']?.trim() || '',
            investor_type: data['Investor type']?.trim() || '',
            min_check_size: cleanCurrency(data['First cheque minimum']),
            max_check_size: cleanCurrency(data['First cheque maximum']),
        };
        results.push(mapped);
    })
    .on('end', async () => {
        console.log(`Read ${results.length} valid records. Deduping...`);

        // Simple dedupe
        const uniqueResults = [];
        const seen = new Set();
        for (const item of results) {
            const id = `${item.firm_name}-${item.website}`.toLowerCase();
            if (!seen.has(id)) {
                seen.add(id);
                uniqueResults.push(item);
            }
        }

        console.log(`Seeding ${uniqueResults.length} unique records into Supabase...`);

        const chunkSize = 50;
        let successCount = 0;
        for (let i = 0; i < uniqueResults.length; i += chunkSize) {
            const chunk = uniqueResults.slice(i, i + chunkSize);
            const { data, error } = await supabase
                .from('european_investors')
                .insert(chunk)
                .select();

            if (error) {
                console.error(`Error in chunk ${Math.floor(i / chunkSize)}:`, error.message, error.details);
            } else {
                successCount += chunk.length;
                console.log(`Inserted chunk ${Math.floor(i / chunkSize) + 1} / ${Math.ceil(uniqueResults.length / chunkSize)} (${successCount} total)`);
            }
        }

        console.log(`Seeding completed! Successfully inserted ${successCount} records.`);
    });
